client/cmd/job/plan.go

```diff
@@ -0,0 +1,218 @@
+	if err != nil {
+		return err
+	}
+
+	switch p.gitProvider {
+	case "gitlab":
+		p.repository, err = gitlab.NewAPI(p.gitURL, p.gitToken)
+	case "github":
+		p.repository, err = github.NewAPI(p.gitURL, p.gitToken)
+	default:
+		return errors.New("unsupported git provider, we currently only support: [github,gitlab]")
+	}
+	if err != nil {
+		return err
+	}
+
+	p.specReadWriter, err = specio.NewJobSpecReadWriter(afero.NewOsFs())
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (p *planCommand) RunE(_ *cobra.Command, _ []string) error {
+	ctx := context.Background()
+	diffs, err := p.repository.CompareDiff(ctx, p.projectID, p.sourceRef, p.destinationRef)
+	if err != nil {
+		return err
+	}
+
+	directories := providermodel.Diffs(diffs).GetAllDirectories(p.appendDirectory)
+	plans := make(plan.Plans, 0, len(directories))
+	p.logger.Info("job plan found changed in directories: %+v", directories)
+
+	for _, directory := range directories {
+		var jobPlan *plan.Plan
+		jobPlan, err = p.describePlanFromDirectory(ctx, directory)
+		if err != nil {
+			return err
+		}
+		if p.verbose {
+			p.logger.Info("[%s] plan operation %s for job %s", jobPlan.NamespaceName, jobPlan.Operation, jobPlan.KindName)
+		}
+		plans = append(plans, jobPlan)
+	}
+
+	sort.SliceStable(plans, plans.SortByOperationPriority)
+	return p.saveFile(plans)
+}
+
+func (p *planCommand) describePlanFromDirectory(ctx context.Context, directory string) (jobPlan *plan.Plan, err error) {
+	var (
+		namespaceName             string
+		sourceRaw, destinationRaw []byte
+		fileName                  = filepath.Join(directory, jobFileName)
+	)
+
+	for _, namespace := range p.clientConfig.Namespaces {
+		if strings.HasPrefix(directory, namespace.Job.Path) {
+			namespaceName = namespace.Name
+			break
+		}
+	}
+	if namespaceName == "" {
+		return nil, fmt.Errorf("failed to find namespace specified on directory: %s", directory)
+	}
+
+	destinationRaw, err = p.repository.GetFileContent(ctx, p.projectID, p.destinationRef, fileName)
+	if err != nil {
+		return nil, errors.Join(err, fmt.Errorf("failed to get file with ref: %s and directory %s", p.destinationRef, directory))
+	}
+	sourceRaw, err = p.repository.GetFileContent(ctx, p.projectID, p.sourceRef, fileName)
+	if err != nil {
+		return nil, errors.Join(err, fmt.Errorf("failed to get file with ref: %s and directory %s", p.sourceRef, directory))
+	}
+
+	var sourceSpec, destinationSpec model.JobSpec
+	if err = yaml.Unmarshal(sourceRaw, &sourceSpec); err != nil {
+		return nil, errors.Join(err, errors.New("failed to unmarshal source job specification"))
+	}
+	if err = yaml.Unmarshal(destinationRaw, &destinationSpec); err != nil {
+		return nil, errors.Join(err, errors.New("failed to unmarshal destination job specification"))
+	}
+
+	jobPlan = &plan.Plan{ProjectName: p.clientConfig.Project.Name, NamespaceName: namespaceName, Kind: plan.KindJob}
+	if len(sourceSpec.Name) == 0 && len(destinationSpec.Name) > 0 {
```

Comment:
want to understand this logic better

Reply:
updated the implementation using method as logic description